internal/chain/executor_query.go

```diff
@@ -223,6 +224,84 @@ func (m *Executor) Query(q *query.Query) (k, v []byte, err *protocol.Error) {
 		if err != nil {
 			return nil, nil, &protocol.Error{Code: protocol.CodeMarshallingError, Message: fmt.Errorf("%v, on Chain %x", err, chr.ChainId)}
 		}
+	case types.QueryTypeKeyPageIndices:
+		chr := query.RequestByUrlAndKey{}
+		err := chr.UnmarshalBinary(q.Content)
+		if err != nil {
+			return nil, nil, &protocol.Error{Code: protocol.CodeUnMarshallingError, Message: err}
+		}
+		u, err := url.Parse(*chr.Url.AsString())
+		if err != nil {
+			return nil, nil, &protocol.Error{Code: protocol.CodeInvalidURL, Message: fmt.Errorf("invalid URL in query %s", chr.Url)}
+		}
+		obj, err := m.queryByChainId(u.ResourceChain())
+		if err != nil {
+			return nil, nil, &protocol.Error{Code: protocol.CodeAddKeyPageIndices, Message: err}
+		}
+		chainHeader := new(state.ChainHeader)
+		if err = obj.As(chainHeader); err != nil {
+			return nil, nil, &protocol.Error{Code: protocol.CodeMarshallingError, Message: fmt.Errorf("inavid object error")}
+		}
```

Comment:
If the record is something other than a key book (or page?), then we need to check the record's key book.

```suggestion
		}
		if chainHeader.Type != types.ChainTypeKeyBook {
			obj, err = m.queryByChainId(chainHeader.KeyBook)
			if err != nil {
				return nil, nil, &protocol.Error{Code: protocol.CodeAddKeyPageIndices, Message: err}
			}
			err = obj.As(chainHeader)
			if err = obj.As(chainHeader); err != nil {
				return nil, nil, &protocol.Error{Code: protocol.CodeMarshallingError, Message: fmt.Errorf("inavid object error")}
			}
		}
```